src/NHibernate/Loader/Hql/QueryLoader.cs

```diff
@@ -397,6 +397,14 @@ private void CheckQuery(QueryParameters queryParameters)
 			{
 				throw new QueryException("ResultTransformer is not allowed for 'select new' queries.");
 			}
+
+			if (Factory.Settings.IsQueryCacheEnabled && queryParameters.Cacheable)
+			{
+				foreach (var persister in _entityPersisters.Where(x => !x.HasUpdateTimestampsCache))
```

Comment:
`_entityPersisters` does not contains all query spaces, for example:
```
var items = s.Query<Order>()
	.Where(o => o.OrderLines.Any())
	.WithOptions(o => o.SetCacheable(true))
	.ToList();
```
for the above query `_entityPersisters` will contain only the persister for `Order` and not for `OrderLine`. I would be nice to have a test for it.

Reply:
I need a table name - persister dictionary for checking persister SupportsQueryCache or not when I look at `BulkOperationCleanupAction(ISessionImplementor session, ISet<string> querySpaces)` constructor its uses factory.GetAllClassMetadata() but I dont want to iterate for all queries, is it a good idea adding SessionFactoryImpl dictionary that contains tableName-persister.

Also I think its not a good idea iterating in `BulkOperationCleanupAction` in every transaction with this new feature I could refactor this class.

```c#
GetTablePersisters()
{
     get{ return tablePersisters;} tablePersisters //initialized in constructor just stores reference of tableName-persister
}
```

Reply:
When we add these functions is this will be breaking change, if its when we will publish these feature?
UpdateTimestampsCache lock also removed as you know @maca88, when you will publish this change.

> with this we can also remove the ICacheableExecutable class as it wouldn't be needed anymore.

I think ICacheableExecutable did not be removed, seperation of logic(cacheable or not) stays inside of IExecuable not ActionQueue.

Reply:
As described below, we can avoid the breaking changes by using extensions methods.

Ok then lets keep `ICacheableExecutable`.